src/serac/numerics/equation_solver.cpp

```diff
@@ -317,13 +344,76 @@ class TrustRegion : public mfem::NewtonSolver {
 #endif
 
   /// finds tau s.t. (z + tau*d)^2 = trSize^2
-  void projectToBoundaryWithCoefs(mfem::Vector& z, const mfem::Vector& d, double trSize, double zz, double zd,
+  void projectToBoundaryWithCoefs(mfem::Vector& z, const mfem::Vector& d, double delta, double zz, double zd,
                                   double dd) const
   {
-    double tau = (std::sqrt((trSize * trSize - zz) * dd + zd * zd) - zd) / dd;
+    // find z + tau d
+    double deltadelta_m_zz = delta * delta - zz;
+    if (deltadelta_m_zz == 0) return;  // already on boundary
+    double tau = (std::sqrt(deltadelta_m_zz * dd + zd * zd) - zd) / dd;
     z.Add(tau, d);
   }
 
+  /// solve the exact trust-region subspace problem with directions ds, and the leftmosts
+  template <typename HessVecFunc>
+  void solveTheSubspaceProblem([[maybe_unused]] mfem::Vector& z, [[maybe_unused]] const HessVecFunc& hess_vec_func,
+                               [[maybe_unused]] const std::vector<const mfem::Vector*> ds,
+                               [[maybe_unused]] const std::vector<const mfem::Vector*> Hds,
+                               [[maybe_unused]] const mfem::Vector& g, [[maybe_unused]] double delta,
+                               [[maybe_unused]] int num_leftmost) const
+  {
+#ifdef SERAC_USE_SLEPC
+    SERAC_MARK_FUNCTION;
+    ++num_subspace_solves;
+
+    std::vector<const mfem::Vector*> directions;
+    for (auto& d : ds) {
+      directions.emplace_back(d);
+    }
+    for (auto& left : left_mosts) {
+      directions.emplace_back(left.get());
+    }
+
+    std::vector<const mfem::Vector*> H_directions;
+    for (auto& Hd : Hds) {
+      H_directions.emplace_back(Hd);
+    }
+    for (auto& H_left : H_left_mosts) {
+      H_directions.emplace_back(H_left.get());
+    }
+
+    std::tie(directions, H_directions) = removeDependentDirections(directions, H_directions);
+
+    mfem::Vector b(g);
+    b *= -1;
+    auto [sol, leftvecs, leftvals, energy_change] =
+        solveSubspaceProblem(directions, H_directions, b, delta, num_leftmost);
+
+    left_mosts.clear();
+    for (auto& lv : leftvecs) {
+      left_mosts.emplace_back(std::move(lv));
+    }
+
+    double base_energy     = computeEnergy(g, hess_vec_func, z);
+    double subspace_energy = computeEnergy(g, hess_vec_func, sol);
+
+    // check that leftmost approximates its eigenvalue
```

Comment:
dead code or needs a note why it staying commented out

Reply:
deleted